!str #16520 #16521 Add ActorRefSource and ActorRefSink

patriknw opened the pull request:
will add something to reference docs if you think this is the right approach

patriknw committed:
!str #16521 Add ActorRefSink

* also rename the factory for ActorSubscriber props Sink,
  from apply to actorSubscriber

patriknw commented:
Refs #16521

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/Sinks.scala:
```diff
@@ -200,3 +200,24 @@ private[akka] final class PropsSink[In](props: Props, val attributes: OperationA
+
+  override def create(materializer: ActorFlowMaterializerImpl, flowName: String) = {
+    val subscriberRef = materializer.actorOf(
+      TellSinkActor.props(ref, materializer.settings.maxInputBufferSize, onCompleteMessage),
```
hmm, do we really need an actor for this?

patriknw replied:
I used an actor because of cancellation when the destination actor terminates, if we don't want that feature we don't need an actor here

rkuhn commented:
This is certainly the most simple approach, but I fear that on the Sink side this is too simplistic: we discussed dropping things on the Source side, and I think the Sink should also drop things instead of just sending. We could recommend people to configure a bounded mailbox for the recipient, another possibility would be to have an extreme simple work pulling protocol:
~~~scala
object Sink {
  def actorRef[T]: Sink[T, ActorRef]
}

case class Request(n: Long, target: ActorRef)
~~~
Then the Subscriber that this is Materialized to will keep a list of requests and serve them in order, which would even allow multiple actors to pull data out of this stream. This could in principle even stop requesting from upstream when there is no demand …

Is this still simple enough to be done right now?

rkuhn commented:
Ah, another thought that makes this more resilient: allow client actors to set their demand to a specific value instead of using normal additive semantics. This allows them to just unconditionally issue that command after restarts.

patriknw commented:
@rkuhn renamed to ActorRefSink and added documentation. Please take a second look.
Also renamed the factory for ActorSubscriber props Sink, from apply to actorSubscriber.

drewhk commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/ActorRefSinkSpec.scala:
```diff
@@ -0,0 +1,50 @@
+  "A ActorRefSink" must {
+
+    "send the elements to the ActorRef" in {
+      Source(List(1, 2, 3)).runWith(Sink.actorRef(testActor, onCompleteMessage = "done"))
```
in many cases this will be more useful than the subscriberProbe

patriknw replied:
exactly what I was thinking about, then you have the full facilities of TestKit
@2m might be something to keep in mind when shaping up the stream-testkit

drewhk commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/ActorRefSinkSpec.scala:
```diff
@@ -0,0 +1,50 @@
+      expectMsg(1)
+      expectMsg(2)
+      system.stop(fw)
+      autoPublisher.subscription.expectCancellation()
```
so nice!

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/ActorRefSinkActor.scala:
```diff
@@ -0,0 +1,36 @@
+/**
+ * INTERNAL API
+ */
+private[akka] class ActorRefSinkActor(ref: ActorRef, highWatermark: Int, onCompleteMessage: Any) extends ActorSubscriber {
```
it would be interesting to experiment later if this can be implemented via a lightweight synthetic actorRef just like PromiseActorRef.

drewhk commented on akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala:
```diff
@@ -99,6 +91,33 @@ object Sink {
+   * upstream, but there is no back-pressure signal from the destination actor,
+   * i.e. if the actor is not consuming the messages fast enough the mailbox
+   * of the actor will grow. For potentially slow consumer actors it is recommended
+   * to use a bounded mailbox with zero `mailbox-push-timeout-time` or use a rate
```
unfortunately that would lose the completion message, too :(

rkuhn replied:
yes, but the target Actor can still watch this one and thereby know when the stream ended

rkuhn replied:
ah, the Terminated would be lost as well …

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/ActorRefSinkActor.scala:
```diff
@@ -0,0 +1,36 @@
+
+  def receive = {
+    case OnNext(elem)      ⇒ ref ! elem
+    case OnError(cause)    ⇒ ref ! Status.Failure(cause)
```
shouldn't we stop the actor here? Or it will be done by the ActorSubscriber superclass?

patriknw replied:
good catch, thanks

ktoso commented on akka-stream/src/main/scala/akka/stream/impl/Sinks.scala:
```diff
@@ -200,3 +200,24 @@ private[akka] final class PropsSink[In](props: Props, val attributes: OperationA
+  override def create(materializer: ActorFlowMaterializerImpl, flowName: String) = {
+    val subscriberRef = materializer.actorOf(
+      ActorRefSinkActor.props(ref, materializer.settings.maxInputBufferSize, onCompleteMessage),
+      name = s"$flowName-tell")
```
`$flowName-actorRef`

patriknw replied:
thx

ktoso commented on akka-stream/src/main/scala/akka/stream/impl/ActorRefSinkActor.scala:
```diff
@@ -0,0 +1,36 @@
+package akka.stream.impl
```
missing copyright

patriknw committed:
!str #16520 Add ActorRefSource

* and rename the factory for ActorPublisherSource,
  from Source.apply to Source.actorPublisher

* including internal buffer, with OverflowStrategy

* support to complete/fail stream

patriknw commented:
I have added one commit with the review feedback.
Another commit with the ActorRefSource, will add reference docs for ActorRefSource when we have sorted out the buffering questions.

patriknw commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/ActorRefSourceSpec.scala:
```diff
@@ -0,0 +1,55 @@
+    "be useful together with buffer" in {
+      val s = StreamTestKit.SubscriberProbe[Int]()
+      // FIXME isn't there a race when starting up, if I send to the ref before the buffer has requested?
+      //       Do we ignore that and fix that with fusing, or shall I add an internal buffer to the ActorRefSourceActor?
```
@drewhk I would like to hear what you think about this

drewhk replied:
This is exactly what I said in the meeting. Just slapping a buffer stage after an element will not fully solve the buffering issue. Fusing is not necessarily possible, in this case for example the source is backed by a custom actor.

Originally I imagined a simple fire-and-forget interface that wraps a bounded buffer and an AtomicLong which tracks pending demand, using only the Publisher interfaces to the downstream, and a simple fire(elem) call. I guess you can put the buffer to the actor, but of course then there is the implicit additional buffer, its mailbox.

patriknw replied:
That matches my conclusions and sounds like adding an internal buffer would not be harmful. Then the buffer size is also very explicit. Opinions @rkuhn

drewhk replied:
I guess the MaterializedValuePublisher I implemented in the other PR can be taken as a template, augmented with a buffer and proper request tracking to have a simple Fire-and-Forget Publisher if we want to pursue that path.

patriknw commented:
I have added the internal buffer. Used same overflow strategies (except Backpressure) as in the buffer stage.

patriknw commented:
Refs #16520

2m commented on akka-stream/src/main/scala/akka/stream/impl/ActorRefSourceActor.scala:
```diff
@@ -0,0 +1,65 @@
+
+  def receive = {
+    case _: Request ⇒
+      // count is tracked by super
```
s/count/totalDemand

ktoso commented on akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala:
```diff
@@ -322,4 +314,33 @@ object Source extends SourceApply {
+   *
+   * The buffer can be disabled by using `bufferSize` of 0 and then received messages are dropped
+   * if there is no demand from downstream. When `bufferSize` is 0 the `overflowStrategy` is does
+   * not matter.
```
not needed `is` in the above sentence?

```
When `bufferSize` is 0 the `overflowStrategy` does not matter.
```
?

ktoso commented on akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:
```diff
@@ -160,6 +152,32 @@ object Source {
+   * there is no space available in the buffer.
+   *
+   * The buffer can be disabled by using `bufferSize` of 0 and then received messages are dropped
+   * if there is no demand from downstream. When `bufferSize` is 0 the `overflowStrategy` is does
```
additional `is` not needed here

ktoso commented:
LGTM, will be very useful!

rkuhn commented on akka-docs-dev/rst/java/stream-integrations.rst:
```diff
@@ -24,6 +29,30 @@ Akka Streams :class:`Source` or :class:`Sink`.
+   There is no back-pressure signal from the destination actor, i.e. if the actor is not consuming
+   the messages fast enough the mailbox of the actor will grow. For potentially slow consumer actors
+   it is recommended to use a bounded mailbox with zero `mailbox-push-timeout-time` or use a rate
+   limiting stage in front of this stage.
```
“… in front of this _Sink_.”

rkuhn commented on akka-stream/src/main/scala/akka/stream/OverflowStrategy.scala:
```diff
@@ -54,6 +54,8 @@ object OverflowStrategy {
    */
   def dropBuffer: OverflowStrategy = DropBuffer
 
+  // FIXME shouldn't there be a strategy that drops the new element?
```
Yes, definitely! I thought that was dropTail, but I was mistaken, please add a ticket.

patriknw replied:
ticket added: #17167

rkuhn commented on akka-stream/src/main/scala/akka/stream/impl/ActorRefSourceActor.scala:
```diff
@@ -0,0 +1,65 @@
+  import akka.stream.actor.ActorPublisherMessage._
+  import akka.stream.OverflowStrategy._
+
+  require(overflowStrategy != OverflowStrategy.Backpressure, "Backpressure overflowStrategy not supported")
```
better move this `require` into the Props factory method

patriknw replied:
ok

rkuhn commented:
LGTM: I agree that this is nice and simple and will be very useful—with one addition, namely how can the ActorRefSource be complete/fail the stream?

patriknw commented:
@rkuhn good point regarding ActorRefSource complete/fail. I suggest that `akka.actor.Status.Failure` can be sent to the ref to fail the stream and `PoisonPill` (or `system.stop`) will complete the stream successfully. WDYT?

rkuhn commented:
Why not make proper message types in the ActorRefSource companion?

patriknw commented:
Because then I will be responsible for making proper serialization for these messages, unless we continue to use java serialization.
I have already used `Status.Failure` for the ActorRefSink so using it on this side also makes it symmetrical.

drewhk commented:
It makes sense for me. And the rest is LGTM

patriknw commented:
ok, I'll complete this now, so @2m can use it

patriknw commented:
I have implemented and documented the completion/fail: c0185b2

Will squash and merge after pr validation

patriknw commented:
failure was unrelated, created ticket: https://github.com/akka/akka/issues/17172

patriknw commented:
rebased and squashed

drewhk commented:
LGTM, merge it!

patriknw commented:
yes, waiting for another validation, had some merge conflicts when rebasing

patriknw commented:
now it was a real test failure, because FixedSizeBuffer had changed (to the better)
one more time!
